File: 5_software_and_its_engineering/1_software_notations_and_tools/common_programming_problems/3_stateful_model/arrays/MinimumSwaps.scala
/*
    Objective:
        Return minimum swaps required

    You are allowex to swap any elements

    Assumptions
        Number always start from 1
 */

/*
    Foreach value, currentIndex pair
        expectedIndexValue = index + 1
        if expectedIndexValue == value, continue
        else
            currentValueCorrectIndex = value - 1
            swap(currentIndex, currentValueCorrectIndex)
 */

def minimumSwaps(list: Array[Int]): Unit = {
  var swaps = 0
  var swapped = true
  val zeroIndexOffset = 1

  def swap(indexA: Int, indexB: Int) = {
    val tempB = list(indexB)
    list(indexB) = list(indexA)
    list(indexA) = tempB
  }
  var i = 1
  while (i > 0) {
    swapped = false
    list.zipWithIndex.foreach { case (value: Int, index: Int) =>
      val expectedIndexValue = index + zeroIndexOffset
      list.foreach(print)
      if (expectedIndexValue != value) {
        swapped = true
        swaps += 1
        val currentValueCorrectIndex = value - zeroIndexOffset
        swap(index, currentValueCorrectIndex)
        println
      }
    }
    i -= 1
    println("out")
  }

  println(s"swaps $swaps")
}

minimumSwaps(Array(4, 3, 1, 2))

File: 5_software_and_its_engineering/0_software_notations_and_tools/unified_discipline_of_computer_programming/programs_per_computational_model/3_stateful/arrays_and_strings/MinimumSwaps.py
"""
    Foreach value, currentIndex pair
        expectedIndexValue = index + 1
        if expectedIndexValue == value, continue
        else 
            currentValueCorrectIndex = value - 1
            swap(currentIndex, currentValueCorrectIndex)

"""


def minimumSwaps(arr):
    def swap(indexA, indexB):
        tempB = arr[indexB]
        arr[indexB] = arr[indexA]
        arr[indexA] = tempB

    swaps = 0
    swapped = True
    zeroIndexOffset = 1
    while swapped:
        swapped = False
        for x in range(0, len(arr)):
            expectedIndexValue = x + zeroIndexOffset
            value = arr[x]
            if expectedIndexValue != value:
                swapped = True
                swaps += 1
                currentValueCorrectIndex = value - zeroIndexOffset
                swap(x, currentValueCorrectIndex)
    return swaps


print(minimumSwaps([4, 3, 1, 2]))
